src/StorageClient.ts

```diff
@@ -39,97 +45,132 @@ export class StorageClient {
   }
 
   private async requestRetry(
-    request: () => Promise<AxiosResponse>,
-    { maxAttempts = 3, delay = 100, onResponse }: RequestOptions = {},
+    requestConfig: StreamBasedRequestConfig, retryOptions: RetryOptions
   ): Promise<AxiosResponse> {
-    let attempts = 0;
+    const { retryDelay, retriesCount, requestTimeout } = validateRetryOptions(retryOptions);
+    let currentRetries = 0;
     let res;
     let err;
-    while (attempts < maxAttempts) {
+    while (currentRetries <= retriesCount) {
       err = null;
       res = null;
-      attempts++;
       try {
-        res = await request();
+        const { axiosReqConfig, getFreshStream } = requestConfig;
+        let bodyAsStream;
+        if (getFreshStream) {
+          const { mime, stream } = await getStreamWithContentType(getFreshStream);
+          axiosReqConfig.headers['content-type'] = mime;
+          bodyAsStream = stream;
+          if (process.env.NODE_ENV === 'test') {
+            log.debug(bodyAsStream); // to insure it's new stream on each call (in tests only)
+          }
+        }
+        if (process.env.NODE_ENV === 'test') {
+          log.trace(retryDelay, retriesCount, requestTimeout);
+        }
+        res = await this.api.request({ ...axiosReqConfig, data: bodyAsStream, timeout: requestTimeout });
       } catch (e) {
+        if (e instanceof ObjectStorageClientError) {
+          throw e;
+        }
         err = e;
+        if (err?.response?.status < 500) throw e; // The request was made and the server responded with a status code
```

Comment:
Just use "validateStatus: null", because you validate it manually, you don't need the help of axios.